moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java

```diff
@@ -77,6 +77,10 @@
   }
 
   @Override public JsonWriter beginArray() throws IOException {
+    if (promoteValueToName) {
+      throw new IllegalStateException(
+          "Array cannot be used as a map key in JSON at path " + getPath());
+    }
     writeDeferredName();
```

Comment:
could be `writeDeferredName("Array")` and writeDeferredName do the check, but then it felt weird to pass in a never used parameter from the callers in value(String) and value(Number).